File: 004-MedianOfTwoSortedArrays/MedianOfTwoSortedArrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: list[int], nums2: list[int]) -> float:
        if(nums1.__len__() > nums2.__len__()):
            temp = nums1
            nums1 = nums2
            nums2 = temp
        total: int = nums1.__len__() + nums2.__len__() 
        half: int = (nums1.__len__() + nums2.__len__() + 1) // 2;
        left: int = 0
        right: int = nums1.__len__()
        while left <= right:
            i = (left + right) // 2;
            j = half - i
            
            l1 = float('-inf') if i == 0 else nums1[i-1]
            r1 = float('inf') if i == nums1.__len__() else nums1[i]
            l2 = float('-inf') if j == 0 else nums2[j-1]
            r2 = float('inf') if j == nums2.__len__() else nums2[j]
            
            if l1 <= r2 and l2 <= r1:
                if total % 2 == 0:
                    return (max(l1, l2) + min(r1, r2)) / 2
                else:
                    return float(max(l1, l2))
            
            if l1 > r2:
                right -= 1
            if l2 > r1:
                left += 1
        return float()

File: 004-MedianOfTwoSortedArrays/MedianOfTwoSortedArrays.java
public class MedianOfTwoSortedArrays {
    public class Solution {
        public double findMedianSortedArrays(int[] nums1, int[] nums2) {
            double median = 0;
            if(nums1.length > nums2.length){
                int[] temp = nums1;
                nums1 = nums2;
                nums2 = temp;
            }
            int total = nums1.length + nums2.length;
            int half = (nums1.length + nums2.length + 1) / 2;
            int left = 0, right = nums1.length;
            while(left <= right){
                int i = (left + right) / 2;
                int j = half - i;

                int l1 = i == 0 ? Integer.MIN_VALUE : nums1[i-1];
                int r1 = i == nums1.length ? Integer.MAX_VALUE : nums1[i];
                int l2 = j == 0 ? Integer.MIN_VALUE : nums2[j - 1];
                int r2 = j == nums2.length ? Integer.MAX_VALUE : nums2[j];

                if(l1 <= r2 && l2 <= r1 ){
                    if(total % 2 == 0){
                        median = ((l1 > l2 ? l1 : l2) +(r1 < r2 ? r1 : r2)) / 2.0;
                    }
                    else {
                        median = l1 > l2 ? l1: l2;
                    }
                    break;
                }
                else if(l1 > r2) right = i - 1;
                else left = i + 1;
            }
            return median;
        }
    }
}